File: src/gx/commands/navigate.py
"""gx up / gx down / gx top / gx bottom — Stack navigation."""

from __future__ import annotations

import typer

from gx.utils.display import print_error, print_info, print_success, print_warning
from gx.utils.git import (
    GitError,
    ensure_git_repo,
    get_current_branch,
    get_head_branch,
    is_clean_working_tree,
    run_git,
)
from gx.utils.stack import get_children, get_parent


def up() -> None:
    """Move up the stack (to child branch)."""
    try:
        ensure_git_repo()
    except GitError as e:
        print_error(str(e))
        raise typer.Exit(1)

    try:
        current = get_current_branch()
    except GitError as e:
        print_error(str(e))
        raise typer.Exit(1)

    children = get_children(current)
    if not children:
        print_info("Already at the top of the stack.")
        raise typer.Exit(0)

    if len(children) > 1:
        print_info(f"Multiple branches stacked on {current}:")
        for child in sorted(children):
            print_info(f"  {child}")
        print_info("Use `gx switch` to pick one.")
        raise typer.Exit(0)

    target = children[0]

    if not is_clean_working_tree():
        print_warning("You have uncommitted changes. They may conflict with the target branch.")

    try:
        run_git(["checkout", target])
        print_success(f"Moved up: {current} -> {target}")
    except GitError as e:
        print_error(f"Failed to switch: {e}")
        raise typer.Exit(1)


def down() -> None:
    """Move down the stack (to parent branch)."""
    try:
        ensure_git_repo()
    except GitError as e:
        print_error(str(e))
        raise typer.Exit(1)

    try:
        current = get_current_branch()
    except GitError as e:
        print_error(str(e))
        raise typer.Exit(1)

    parent = get_parent(current)
    if parent is None:
        print_info(f"{current} is not in the stack. Use `gx switch` to navigate.")
        raise typer.Exit(0)

    if not is_clean_working_tree():
        print_warning("You have uncommitted changes. They may conflict with the target branch.")

    try:
        run_git(["checkout", parent])
    except GitError as e:
        print_error(f"Failed to switch: {e}")
        raise typer.Exit(1)

    head_branch = get_head_branch()
    if parent == head_branch:
        print_info(f"Moved down to {parent} (trunk).")
    else:
        print_success(f"Moved down: {current} -> {parent}")


def top() -> None:
    """Jump to the top of the stack (furthest from trunk)."""
    try:
        ensure_git_repo()
    except GitError as e:
        print_error(str(e))
        raise typer.Exit(1)

    try:
        current = get_current_branch()
    except GitError as e:
        print_error(str(e))
        raise typer.Exit(1)

    target = current
    visited: set[str] = {current}

    while True:
        children = sorted(get_children(target))
        if not children:
            break
        if len(children) > 1:
            print_info(f"Stack branches at {target}. Cannot determine a single top.")
            print_info(f"Children: {', '.join(children)}")
            print_info("Use `gx switch` to pick a specific branch.")
            raise typer.Exit(0)
        next_branch = children[0]
        if next_branch in visited:
            print_warning("Cycle detected in stack config.")
            break
        visited.add(next_branch)
        target = next_branch

    if target == current:
        print_info("Already at the top of the stack.")
        raise typer.Exit(0)

    if not is_clean_working_tree():
        print_warning("You have uncommitted changes. They may conflict with the target branch.")

    try:
        run_git(["checkout", target])
        print_success(f"Jumped to top: {current} -> {target}")
    except GitError as e:
        print_error(f"Failed to switch: {e}")
        raise typer.Exit(1)


def bottom() -> None:
    """Jump to the bottom of the stack (closest to trunk)."""
    try:
        ensure_git_repo()
    except GitError as e:
        print_error(str(e))
        raise typer.Exit(1)

    try:
        current = get_current_branch()
    except GitError as e:
        print_error(str(e))
        raise typer.Exit(1)

    head_branch = get_head_branch()

    # If on trunk or not in stack, try to enter the stack
    parent = get_parent(current)
    if current == head_branch or parent is None:
        children = sorted(get_children(current))
        if not children:
            print_info("No stack branches found.")
            raise typer.Exit(0)
        if len(children) == 1:
            if not is_clean_working_tree():
                print_warning("You have uncommitted changes.")
            try:
                run_git(["checkout", children[0]])
                print_success(f"Jumped to bottom: {current} -> {children[0]}")
            except GitError as e:
                print_error(f"Failed to switch: {e}")
                raise typer.Exit(1)
        else:
            print_info(f"Multiple stacks branching from {current}:")
            for child in children:
                print_info(f"  {child}")
            print_info("Use `gx switch` to pick one.")
        return

    # Walk down toward trunk to find the bottom-most managed branch
    target = current
    visited: set[str] = {current}

    while True:
        p = get_parent(target)
        if p is None or p == head_branch or p not in visited and get_parent(p) is None:
            break
        if p in visited:
            print_warning("Cycle detected in stack config.")
            break
        visited.add(p)
        target = p

    if target == current:
        print_info("Already at the bottom of the stack.")
        raise typer.Exit(0)

    if not is_clean_working_tree():
        print_warning("You have uncommitted changes. They may conflict with the target branch.")

    try:
        run_git(["checkout", target])
        print_success(f"Jumped to bottom: {current} -> {target}")
    except GitError as e:
        print_error(f"Failed to switch: {e}")
        raise typer.Exit(1)

File: cmd/navigate.go
package cmd

import (
	"fmt"

	"github.com/mubbie/gx-cli/internal/git"
	"github.com/mubbie/gx-cli/internal/stack"
	"github.com/mubbie/gx-cli/internal/ui"
	"github.com/spf13/cobra"
)

func init() {
	rootCmd.AddCommand(&cobra.Command{Use: "up", Short: "Move up the stack (to child branch)", RunE: runUp})
	rootCmd.AddCommand(&cobra.Command{Use: "down", Short: "Move down the stack (to parent branch)", RunE: runDown})
	rootCmd.AddCommand(&cobra.Command{Use: "top", Short: "Jump to the top of the stack", RunE: runTop})
	rootCmd.AddCommand(&cobra.Command{Use: "bottom", Short: "Jump to the bottom of the stack", RunE: runBottom})
}

func runUp(cmd *cobra.Command, args []string) error {
	current, err := requireBranch()
	if err != nil {
		return nil
	}
	cfg, err := stack.Load()
	if err != nil {
		ui.PrintError(err.Error())
		return nil
	}
	children := cfg.ChildrenOf(current)
	if len(children) == 0 {
		ui.PrintInfo("Already at the top of the stack.")
		return nil
	}
	if len(children) > 1 {
		ui.PrintInfo(fmt.Sprintf("Multiple branches stacked on %s:", current))
		for _, c := range children {
			ui.PrintInfo("  " + c)
		}
		ui.PrintInfo("Use `gx switch` to pick one.")
		return nil
	}
	warnIfDirty()
	if err := switchTo(children[0]); err != nil {
		return nil
	}
	ui.PrintSuccess(fmt.Sprintf("Moved up: %s -> %s", current, children[0]))
	return nil
}

func runDown(cmd *cobra.Command, args []string) error {
	current, err := requireBranch()
	if err != nil {
		return nil
	}
	cfg, err := stack.Load()
	if err != nil {
		ui.PrintError(err.Error())
		return nil
	}
	parent := cfg.ParentOf(current)
	if parent == "" {
		ui.PrintInfo(fmt.Sprintf("%s is not in the stack. Use `gx switch` to navigate.", current))
		return nil
	}
	warnIfDirty()
	if err := switchTo(parent); err != nil {
		return nil
	}
	if parent == git.HeadBranch() {
		ui.PrintInfo(fmt.Sprintf("Moved down to %s (trunk).", parent))
	} else {
		ui.PrintSuccess(fmt.Sprintf("Moved down: %s -> %s", current, parent))
	}
	return nil
}

func runTop(cmd *cobra.Command, args []string) error {
	current, err := requireBranch()
	if err != nil {
		return nil
	}
	cfg, err := stack.Load()
	if err != nil {
		ui.PrintError(err.Error())
		return nil
	}

	target := current
	visited := map[string]bool{current: true}
	for {
		children := cfg.ChildrenOf(target)
		if len(children) == 0 {
			break
		}
		if len(children) > 1 {
			ui.PrintInfo(fmt.Sprintf("Stack branches at %s. Cannot determine a single top.", target))
			ui.PrintInfo("Use `gx switch` to pick a specific branch.")
			return nil
		}
		if visited[children[0]] {
			ui.PrintWarning("Cycle detected in stack config.")
			break
		}
		visited[children[0]] = true
		target = children[0]
	}

	if target == current {
		ui.PrintInfo("Already at the top of the stack.")
		return nil
	}
	warnIfDirty()
	if err := switchTo(target); err != nil {
		return nil
	}
	ui.PrintSuccess(fmt.Sprintf("Jumped to top: %s -> %s", current, target))
	return nil
}

func runBottom(cmd *cobra.Command, args []string) error {
	current, err := requireBranch()
	if err != nil {
		return nil
	}
	cfg, err := stack.Load()
	if err != nil {
		ui.PrintError(err.Error())
		return nil
	}

	head := git.HeadBranch()
	parent := cfg.ParentOf(current)

	// On trunk or not in stack: enter the stack
	if current == head || parent == "" {
		children := cfg.ChildrenOf(current)
		if len(children) == 0 {
			ui.PrintInfo("No stack branches found.")
			return nil
		}
		if len(children) == 1 {
			warnIfDirty()
			if err := switchTo(children[0]); err != nil {
				return nil
			}
			ui.PrintSuccess(fmt.Sprintf("Jumped to bottom: %s -> %s", current, children[0]))
		} else {
			ui.PrintInfo(fmt.Sprintf("Multiple stacks branching from %s:", current))
			for _, c := range children {
				ui.PrintInfo("  " + c)
			}
			ui.PrintInfo("Use `gx switch` to pick one.")
		}
		return nil
	}

	// Walk down toward trunk
	target := current
	visited := map[string]bool{current: true}
	for {
		p := cfg.ParentOf(target)
		if p == "" || p == head {
			break
		}
		if visited[p] {
			ui.PrintWarning("Cycle detected in stack config.")
			break
		}
		visited[p] = true
		if cfg.ParentOf(p) == "" || cfg.ParentOf(p) == head {
			target = p
			break
		}
		target = p
	}

	if target == current {
		ui.PrintInfo("Already at the bottom of the stack.")
		return nil
	}
	warnIfDirty()
	if err := switchTo(target); err != nil {
		return nil
	}
	ui.PrintSuccess(fmt.Sprintf("Jumped to bottom: %s -> %s", current, target))
	return nil
}
